core/lib/common/worker.js

```diff
@@ -149,68 +149,60 @@ module.exports = class Worker {
 		await retry(
 			async () => {
 				attempt++;
-				this.logger.log(`Preparing to flash, attempt ${attempt}...`);
-
-				await new Promise(async (resolve, reject) => {
-					const req = rp.post({ uri: `${this.url}/dut/flash`, timeout: 0 });
-
-					req.catch((error) => {
-						this.logger.log(`client side error: `)
-						this.logger.log(error.message)
-						reject(error);
-					});
-					req.finally(() => {
-						if (lastStatus !== 'done') {
-							reject(new Error('Unexpected end of TCP connection'));
-						}
-
-						resolve();
-					});
-
-					let lastStatus;
-					req.on('data', (data) => {
-						const computedLine = RegExp('(.+?): (.*)').exec(data.toString());
-
-						if (computedLine) {
-							if (computedLine[1] === 'error') {
-								req.cancel();
-								reject(new Error(computedLine[2]));
-							}
-
-							if (computedLine[1] === 'progress') {
-								once(() => {
-									this.logger.log('Flashing');
-								});
-								// Hide any errors as the lines we get can be half written
-								const state = JSON.parse(computedLine[2]);
-								if (state != null && isNumber(state.percentage)) {
-									this.logger.status({
-										message: 'Flashing',
-										percentage: state.percentage,
-									});
-								}
-							}
-
-							if (computedLine[1] === 'status') {
-								lastStatus = computedLine[2];
-							}
-						}
-					});
-
-					pipeline(
+				this.logger.log(`Sending image to worker, attempt ${attempt}...`);
+				// Send image to worker
+				const req = rp.post({ uri: `${this.url}/dut/sendImage`, timeout: 0 });
+
+				// if sending fails, will recieve a 500 code from the worker so should reject
+				try {
+					await pipeline(
 						fs.createReadStream(imagePath),
 						createGzip({ level: 6 }),
 						req,
 					);
-				});
-				this.logger.log('Flash completed');
+				} catch (e) {
+					throw new Error(e.message)
+				}
 			},
 			{
 				max_tries: 5,
 				interval: 1000 * 5,
 				throw_original: true,
 			},
 		);
+
+		this.logger.log(`Image sent to worker successfully`);
+
+		// Initiate flashing process on the worker
+		await doRequest({ method: 'POST', uri: `${this.url}/dut/flashImage`});
+
+		// Now wait for worker to flash DUT - poll the worker to check
+		await retry(
+			async () => {
+				this.logger.log(`Checking if DUT is flashed...`);
+				// check flashing state
+				let flashState = await doRequest({ uri: `${this.url}/dut/flashState` });
+				
+				switch(flashState){
+					case "DONE":
+						this.logger.log('Flashing completed!');
+						return
+					case "PENDING":
+						this.logger.log('Flashing still pending...');
+						throw new Error('Flashing still pending...');
+					case "ERROR":
+						this.logger.log('Error flashing DUT, retrying...');
+						// if there was an error, retry the flash
+						await doRequest({ method: 'POST', uri: `${this.url}/dut/flashImage`});
+						throw new Error('Error flashing DUT, retrying...');
```

Comment:
Not sure why are are flashing here again. 

For Case ERROR, we logged the error, we throw the error, we retry and with that retry the loop should initiate flashing at the top.

Reply:
Agreed, I see this is only being used as an error case retry rather than retry on every loop. Sounds good to me.